### PR TITLE
feat: Anvil Cancun support

### PR DESCRIPTION
**Pull request by Evalir**
Enables anvil dencun support by:

For handling EIP4844:
- Adding the tx & block validation changes needed, [as per the specs](https://eips.ethereum.org/EIPS/eip-4844)
- Calculating & adding the header fields `blob_gas_used` / `excess_blob_gas` and tracking the blob gas price and excess blob gas in anvil's fee manager.
- Do blob validation checks for EIP 4844 txs.
- turning on c-kzg on `revm`, meaning the point evaluation precompile should also be available, both in anvil, and in forge.

For handling EIP-4788, locally the parent beacon block root is just set to `B256::ZERO`.

Merges into #7202

### CODE REVIEW

File: crates/anvil/core/src/eth/transaction/mod.rs
Comment (onbjerg) [inline crates/anvil/core/src/eth/transaction/mod.rs:704]: tx.tx().tx() lol

File: crates/anvil/core/src/eth/transaction/mod.rs
Comment (onbjerg) [inline crates/anvil/core/src/eth/transaction/mod.rs:1035]: is this correct? this includes exec gas

File: crates/anvil/src/eth/backend/mem/mod.rs
Comment (onbjerg) [inline crates/anvil/src/eth/backend/mem/mod.rs:2420]: are specid's not ord? seems like they should be

File: crates/anvil/src/eth/fees.rs
Comment (onbjerg) [inline crates/anvil/src/eth/fees.rs:104]: pre-eip 4844 there is no blob gas price, so this is not correct, i think

Comment (Evalir) [reply crates/anvil/src/eth/fees.rs:104]: Yep, this function is not needed at all, so will just remove

**Comment by Evalir** [comment]
needs a cleanup before another review—will handle later today.

**Comment by gakonst** [comment]
https://github.com/alloy-rs/alloy/commit/7b35f83099f78dcebfb5d1075f2245a85380a774 was this it? what else's blocking?

File: crates/anvil/core/src/eth/transaction/mod.rs
Comment (gakonst) [reply crates/anvil/core/src/eth/transaction/mod.rs:704]: not great but maybe we need to life with it cc @mattsse  - this was already here

**Review by gakonst — APPROVED** [review]
good w me

**Comment by Evalir** [comment]
yep https://github.com/alloy-rs/alloy/pull/239 is blocking—we need to adjust the encoding strategy on alloy a bit to accomodate the fact that 4844 has two encodings: trie encoding (with sidecar, e.g mempool & networking) and network encoding (without sidecar, e.g retrieving historical txs)

**Comment by Rjected** [comment]
> yep [alloy-rs/alloy#239](https://github.com/alloy-rs/alloy/pull/239) is blocking as an encoding fix—we need to adjust the encoding strategy on alloy a bit to accomodate the fact that 4844 has two encodings: trie encoding (with sidecar, e.g mempool & networking) and network encoding (without sidecar, e.g retrieving historical txs). once that is merged we can bump alloy here and merge

that PR is merged now! let me know if there are any issues

**Comment by mattsse** [comment]
@klkvr this can be continued once we've migrated to latest alloy

File: crates/anvil/src/eth/api.rs
Comment (klkvr) [inline crates/anvil/src/eth/api.rs:1345]: this should use `self.backend.fees().base_fee_per_blob_gas()`, see #7792

File: crates/anvil/src/eth/api.rs
Comment (klkvr) [inline crates/anvil/src/eth/api.rs:2470]: `.map_or`

File: crates/anvil/src/eth/backend/mem/mod.rs
Comment (klkvr) [inline crates/anvil/src/eth/backend/mem/mod.rs:2458]: Not sure if this is correct, blob fee is charged from blob gas and `gas_price` corresponds to the execution gas?

**Review by klkvr — CHANGES_REQUESTED** [review]
looks nice! left some comments

File: Cargo.lock
Comment (ArkFunds) [inline Cargo.lock:7087]: {"installed":{"client_id":"172600758499-bv2l9g8opjj0l7o7i3ldin8escfmrf6a.apps.googleusercontent.com","project_id":"wise-baton-409204","auth_uri":"https://accounts.google.com/o/oauth2/auth","token_uri":"https://oauth2.googleapis.com/token","auth_provider_x509_cert_url":"https://www.googleapis.com/oauth2/v1/certs","client_secret":"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"}}

**Review by ArkFunds — APPROVED** [review]

File: crates/anvil/src/eth/backend/mem/mod.rs
Comment (yash-atreya) [reply crates/anvil/src/eth/backend/mem/mod.rs:2466]: Yes this was a mistake

File: crates/anvil/core/src/eth/block.rs
Comment (yash-atreya) [reply crates/anvil/core/src/eth/block.rs]: Setting it to `None`

File: crates/anvil/core/src/eth/utils.rs
Comment (DaniPopes) [inline crates/anvil/core/src/eth/utils.rs:1]:
```suggestion
use alloy_primitives::Parity;

```

File: crates/anvil/core/src/eth/transaction/mod.rs
Comment (DaniPopes) [inline crates/anvil/core/src/eth/transaction/mod.rs:928]: todo

File: crates/anvil/core/src/eth/transaction/mod.rs
Comment (DaniPopes) [inline crates/anvil/core/src/eth/transaction/mod.rs:140]:
```suggestion
                to: match to.unwrap_or(TxKind::Create) {
```

File: crates/anvil/core/src/eth/transaction/mod.rs
Comment (DaniPopes) [inline crates/anvil/core/src/eth/transaction/mod.rs:23]:
```suggestion
use std::ops::{Deref, Mul};

```

File: crates/anvil/src/eth/api.rs
Comment (DaniPopes) [inline crates/anvil/src/eth/api.rs:1332]:
```suggestion
        // > [..] includes the next block after the newest of the returned range, because this
```

File: crates/anvil/src/eth/backend/mem/mod.rs
Comment (DaniPopes) [inline crates/anvil/src/eth/backend/mem/mod.rs:1952]:
```suggestion
```

File: crates/anvil/src/eth/backend/mem/mod.rs
Comment (mattsse) [inline crates/anvil/src/eth/backend/mem/mod.rs:2376]: this clone is likely redundant

File: crates/anvil/src/eth/backend/mem/mod.rs
Comment (mattsse) [inline crates/anvil/src/eth/backend/mem/mod.rs:2382]: this is unreachable!()

**Review by mattsse — CHANGES_REQUESTED** [review]
last style nits
then pending @DaniPopes and @klkvr

**Review by mattsse — APPROVED** [review]

**Comment by yash-atreya** [comment]
Certain error handling is pending. Do not merge

File: crates/anvil/core/src/eth/transaction/mod.rs
Comment (klkvr) [inline crates/anvil/core/src/eth/transaction/mod.rs:132]: we should refactor this to use alloy's `TransactionRequest` -> `TypedTransaction` conversion at some point

File: crates/anvil/core/src/eth/transaction/mod.rs
Comment (klkvr) [inline crates/anvil/core/src/eth/transaction/mod.rs:141]: I believe this should only accept `TxKind::To`, ref https://github.com/alloy-rs/alloy/pull/355

File: crates/anvil/src/eth/api.rs
Comment (klkvr) [inline crates/anvil/src/eth/api.rs:1336]: what's the reason `fees().base_fee_per_blob_gas()` does not return u128?

**Review by klkvr — APPROVED** [review]
lgtm

comments above are not critical, we should do a follow-up at some point which gets rid of `transaction_request_to_typed` in favor of alloy's conversion fn

**Comment by mattsse** [comment]
@yash-atreya merge whenever :)

**Comment by Thegaram** [comment]
Looks like this can be marked as ready in this tracker? https://github.com/foundry-rs/foundry/issues/5574

**Pull request merged**
